lib/rules/no-mutation-props.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * @fileoverview Prevent direct mutation of this.props
+ * @author Ian Schmitz
+ */
+'use strict';
+
+var has = require('has');
+var Components = require('../util/Components');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Prevent direct mutation of this.props',
+      category: 'Possible Errors',
+      recommended: true
+    }
+  },
+
+  create: Components.detect(function (context, components, utils) {
+
+    /**
+     * Checks if the component is valid
+     * @param {Object} component The component to process
+     * @returns {Boolean} True if the component is valid, false if not.
+     */
+    function isValid(component) {
+      return Boolean(component && !component.mutateProps);
+    }
+
+    /**
+     * Reports this.props mutations for a given component
+     * @param {Object} component The component to process
+     */
+    function reportMutations(component) {
+      var mutation;
+      for (var i = 0, j = component.mutations.length; i < j; i++) {
+        mutation = component.mutations[i];
+        context.report({
+          node: mutation,
+          message: 'Do not mutate props.'
+        });
+      }
+    }
+
+    // --------------------------------------------------------------------------
+    // Public
+    // --------------------------------------------------------------------------
+
+    return {
+      AssignmentExpression: function (node) {
+        var item;
+
+        if (!node.left || !node.left.object || !node.left.object.object) {
+          return;
+        }
+
+        item = node.left.object;
+        while (item.object.property) {
+          item = item.object;
+        }
+
+        if (item.object.type === 'ThisExpression' && item.property.name === 'props') {
+          var component = components.get(utils.getParentComponent());
+          var mutations = component && component.mutations || [];
+          mutations.push(node.left.object);
+          components.set(node, {
+            mutateProps: true,
+            mutations: mutations
+          });
+        }
+      },
+
+      'Program:exit': function () {
+        var list = components.list();
+        for (var component in list) {
```

Comment:
let's use `Object.keys(list).forEach` if it's an object, and if it's an array, `list.forEach` - `for..in` should be avoided.